module/synchronization/core.go

```diff
@@ -237,31 +254,39 @@ func (c *Core) getRequestStatus(height uint64, blockID flow.Identifier) *Status
 // prune removes any pending requests which we have received and which is below
 // the finalized height, or which we received sufficiently long ago.
 func (c *Core) prune(final *flow.Header) {
+	if c.localFinalizedHeight >= final.Height {
+		return
+	}
+
+	c.localFinalizedHeight = final.Height
 
 	// track how many statuses we are pruning
 	initialHeights := len(c.heights)
 	initialBlockIDs := len(c.blockIDs)
 
-	for height := range c.heights {
+	for height, status := range c.heights {
 		if height <= final.Height {
```

Comment:
Some idea for future optimization: this is a O(N) loop, and later there is another O(N) loop over c.blockIDs. 

Can we do better here?

Reply:
I think if we added a min priority queue and kept our map for looking up blocks by height/id in O(1). We could change these loops to O(mlogn) where m is the number of blocks we actually need to prune which should be small assuming we're slowly increasing the finalized height and n is the number of blocks we're storing. 

We should see the number of blocks we're actually storing with the new metrics and be able to figure out if it would be worth doing.